proposals/0066-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,245 @@
+won’t activate this feature if its corresponding feature account no longer
+exists on-chain.
+This is how the runtime currently handles features and can be observed here:
+<https://github.com/solana-labs/solana/blob/170478924705c9c62dbeb475c5425b68ba61b375/runtime/src/bank.rs#L8119-L8150>.
+
+### Activating Supported Features
+
+Once the epoch concludes, the runtime uses the validator support signals to
+activate the proper features.
+
+To do this, the runtime walks all of the PDAs derived from the vote accounts,
+examines their bit-masks, and sums the stake supporting each feature-gate to
+determine which feature gates meet the desired threshold of supported stake.
+
+For the initial implementation, the threshold will be hard-coded to 95% of
+stake, but future iterations on the Feature Program could allow feature
+key-holders to set a custom threshold for each feature gate
+(including 0%, ie. –yolo).
+
+Once a list of stake-supported features is generated, the runtime rechecks the
+Feature accounts in case any have been revoked. Then the runtime processes the
+activations using the existing feature logic.
+
+At this point, the runtime clears the existing validator support PDA accounts
+and returns to step 1 to generate the next list of queued features.
+
+### Conclusion
+
+The current familiar process of activating a feature with
+`solana feature activate` will now merely suggest this feature be considered
+for activation by the aforementioned mechanism. In other words, it’s no longer
+a guarantee that your feature will be activated when you run this command.
+
+Instead, the runtime will decide through stake whether or not a feature should
+be activated. This will be especially useful in a world where two separate
+clients will aim to push features and want to agree on their activation.
+
+## Alternatives Considered
+
+### Representing the Current Feature Set (Alternative)
+
+An alternate approach to the Feature Gate program’s on-chain storage is to
+instead use one single PDA to store the queue and reset it every epoch. This
+would allow us to use one less account and one less epoch, but it would make
+things trickier when it comes to the timing of validators signaling support for
+features.
+
+For example, a key-holder might submit a feature for activation at a point late
+in the epoch, so validators would need to be able to signal support for this
+feature whenever it is submitted. This means validators would have to
+consistently check the feature queue for changes, and possibly issue more
+signals in the epoch.
+
+This approach may increase the total required work for a node in a given epoch.
+
+### Signaling Support for Features (Alternative)
+
+Rather than sending a transaction to the Feature Gate program, a node could
+instead signal the bit mask in their block header whenever they are the leader.
+
+This would introduce a change to the block header and remove this information
+from being publicly and transparently on-chain.
```

Comment:
True. Should I add this as another alternative?

Reply:
No, modifying the block header is pretty much a non-starter

Reply:
I meant the leader-signaled transaction aspect proposed by @ripatel-fd.

Reply:
@ripatel-fd , our concern with signalling support only in a leader's slot(s) is that it makes validators less flexible as to when they can upgrade and signal support for a new feature. For instance, imagine a node that is leader for only the first 4 slots of the epoch. Once those slots pass, that node must wait at least 431_996 slots to signal support, regardless of whether the operator finds that (~2 days) to be plenty of time to assess the new version/feature set and perform the upgrade.

Reply:
We've relaxed the constraint to prefer "send it sometime in the epoch before 128 slots left". If we still want to discuss the leader approach, (in the interest of time passed since this thread) let's open a new thread if we want to discuss further!